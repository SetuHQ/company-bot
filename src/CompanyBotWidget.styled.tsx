import styled from "styled-components";
import "./styles/fonts.css";

import { lighten, darken, rgba, transparentize } from "polished";
import { defaultColours } from "fictoan-react";

import { setuColours } from "./styles/setuColours";

export const CompanyBotWidgetStyled = styled.div`
    font-family: "Ibm Plex Sans", sans-serif;
    --color-primary: #152533;
    --color-text: white;
    --input-height: 46px;

    .feedback-widget {
        font-family: "Ibm Plex Sans", sans-serif;
        z-index: 99999;
        position: fixed;
        bottom: 40px;
        right: 40px;
        width: 100vw;
        height: 0;
        font-size: 15px;
        line-height: 1.4;

        -webkit-text-size-adjust: 100%;
        -webkit-font-smoothing: antialiased;
        text-rendering: optimizeLegibility;
        -moz-osx-font-smoothing: grayscale;
    }

    .feedback-widget-trigger-button {
        font-family: "Ibm Plex Sans", sans-serif;
        cursor: pointer;
        z-index: 1;
        position: absolute;
        right: 0;
        bottom: 0;
        width: 60px;
        height: 60px;
        background: #c6eff0;
        display: flex;
        align-items: center;
        justify-content: center;
        border-radius: 60px;

        svg {
            height: 100%;
        }
    }

    .widget-popover {
        font-family: "Ibm Plex Sans", sans-serif;
        position: fixed;
        top: 0px;
        right: 0px;
        z-index: 2147483999;
        display: flex;
        background-color: white;
        color: ${lighten(0.16, setuColours.murkyNight)};
        width: 28%;
        height: 100%;
        overflow-y: scroll;
    }

    .back-icon {
        font-family: "Ibm Plex Sans", sans-serif;
        cursor: pointer;
        color: ${lighten(0.16, setuColours.murkyNight)};
    }
    .input-field {
        font-family: "Ibm Plex Sans", sans-serif;
    }
    .input-field:focus,
    .input-field:active {
        border: 2px solid hsl(181, 58%, 53%);
    }
    .submit-button {
        font-family: "Ibm Plex Sans", sans-serif;
        background-color: hsl(181, 58%, 53%);
        border: hsl(181, 58%, 53%);
    }
    @media (max-width: 900px) {
        .widget-popover {
            width: 100%;
        }
    }
`;
