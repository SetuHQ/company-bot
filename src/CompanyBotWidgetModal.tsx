import React, { useState, useEffect, FormEvent, ChangeEvent } from "react";
import {
    Button,
    Element,
    Heading,
    Text,
    FormWrapper,
    InputField,
    Card,
    HRule,
} from "fictoan-react";

import SetuBotPreview from "./components/CompanyBotPreview/CompanyBotPreview";
// import { getPromptResults } from '../services/get-prompt-results';

interface SetuBotWidgetModalProps {
    closeModal: () => void;
    description?: string;
    heading?: string;
    projectKey: string;
}
const SetuBotWidgetModal = ({
    closeModal,
    projectKey,
    heading,
    description,
}: SetuBotWidgetModalProps) => {
    const [question, setQuestion] = useState("");
    const [answer, setAnswer] = useState<string | null | unknown>("");
    const [isLoading, setIsLoading] = useState(false);
    const [error, setError] = useState("");
    const [isSearching, setIsSearching] = useState(false);
    const [references, setReferences] = useState([]);
    const [wasYesClicked, setWasYesClicked] = useState(false);
    const [wasNoClicked, setWasNoClicked] = useState(false);
    const [contextText, setContextText] = useState(
        heading
            ? heading + ` is generating your answer, please wait...`
            : "CompanyBot is generating your answer, please wait..."
    );

    useEffect(() => {
        setTimeout(() => {
            setWasNoClicked(false);
            setWasYesClicked(false);
        }, 5000);
    }, [wasNoClicked, wasYesClicked]);

    // const handleSubmit = (e: FormEvent) => {
    //     e.preventDefault();
    //     setError("");
    //     setIsLoading(true);
    //     setIsSearching(true);
    //     // getPromptResults(projectKey, question)
    //     //   .then((res) => {
    //     //     setAnswer(res[1]);
    //     //     setReferences(res[0]);
    //     //     setIsSearching(false);
    //     //     setIsLoading(false);
    //     //   })
    //     //   .catch((err) => {
    //     //     setError("Sorry, I'm currently not able to answer your question");
    //     //     setIsLoading(false);
    //     //     setIsSearching(false);
    //     //   });
    // };

    const handleSubmit = async (e: FormEvent) => {
        e.preventDefault();
        setError("");
        setIsLoading(true);
        setIsSearching(true);

        try {
            const res = await fetch(
                `https://api.markprompt.com/v1/completions`,
                {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                    },
                    body: JSON.stringify({
                        prompt: question,
                        model: "gpt-4",
                        iDontKnowMessage:
                            "Sorry, LegalBot is currently not able to answer your question",
                        projectKey: "sk_test_6UdTwE0lPnUlpFpEGNJbmoL6Ihwat7BT",
                    }),
                }
            );

            if (!res.ok || !res.body) {
                const text = await res.text();
                console.error("Error:", text);
                setAnswer(
                    "Sorry, LegalBot is currently not able to answer your question"
                );
                return;
            }

            const reader = res.body.getReader();
            const decoder = new TextDecoder();
            let done = false;
            let startText = "";
            let didHandleHeader = false;
            let refs = [];

            setIsSearching(false);

            while (!done) {
                const { value, done: doneReading } = await reader.read();
                done = doneReading;
                const chunkValue = decoder.decode(value);

                if (!didHandleHeader) {
                    startText = startText + chunkValue;
                    if (startText.includes("___START_RESPONSE_STREAM___")) {
                        const parts = startText.split(
                            "___START_RESPONSE_STREAM___"
                        );
                        try {
                            refs = JSON.parse(parts[0]);
                        } catch {}
                        setAnswer((prev: any) => prev + parts[1]);
                        didHandleHeader = true;
                    }
                } else {
                    setAnswer((prev: any) => prev + chunkValue);
                }
            }
            setReferences(refs);
            setIsLoading(false);
            setContextText("Here is what LegalBot came up with:");
        } catch (e) {
            console.error("Error", e);
            setAnswer(
                "Sorry, LegalBot is currently not able to answer your question"
            );
        }

        // initProgess = 0;
        // setProgress({
        //     value: 0,
        //     text: stepsinSetuBot[0],
        // });
        // moveProgressBar();
        // mixpanel.track("SetuBot question", {
        //     question: question,
        // });
        // cmsService
        //     .getPromptResults({ prompt: question })
        //     .then((res) => {
        //         parseMDX(res.answer, res.refs);
        //     })
        //     .catch((err) => {
        //         setError(
        //             "Sorry, SetuBot is currently not able to answer your question"
        //         );
        //         setIsLoading(false);
        //         setIsSearching(false);
        //     });
    };

    return (
        <Card
            className="widget-popover"
            shape="rounded"
            padding="micro"
            shadow="hard"
        >
            <Element as="div" className="full-width" marginBottom="nano">
                <Element
                    as="span"
                    marginRight="nano"
                    className="back-icon"
                    onClick={() => {
                        closeModal();
                    }}
                >
                    ‚Üê Back
                </Element>
                <Heading as="h4" marginBottom="micro" marginTop="micro">
                    Ask {heading || "CompanyBot"}
                    <br />
                    <Text size="micro" marginTop="nano">
                        {description || "Get answers regarding our company"}
                    </Text>
                </Heading>

                <FormWrapper onSubmit={handleSubmit}>
                    <InputField
                        placeholder="Enter any question you need answers for"
                        type="text"
                        className="input-field"
                        value={question}
                        onChange={(e: ChangeEvent<HTMLInputElement>) => {
                            setQuestion(e.target.value);
                            setAnswer("");
                        }}
                    />

                    <Button
                        kind="primary"
                        shape="rounded"
                        shadow="hard"
                        isFullWidth
                        isLoading={isLoading}
                        disabled={question === ""}
                        className="submit-button"
                    >
                        Ask {heading || "CompanyBot"}
                    </Button>
                </FormWrapper>
                {error !== "" && <Text marginTop="micro">{error}</Text>}
                {isSearching ? (
                    <Text marginTop="nano">
                        {heading || "CompanyBot"} is thinking...
                    </Text>
                ) : null}
                {answer && !isSearching ? (
                    <>
                        <Element
                            as="div"
                            marginTop="micro"
                            marginBottom="micro"
                        >
                            <Text weight="600">{contextText}</Text>
                            <br />
                            <SetuBotPreview mdx={answer} />
                        </Element>
                    </>
                ) : null}
            </Element>
        </Card>
    );
};

export default SetuBotWidgetModal;
