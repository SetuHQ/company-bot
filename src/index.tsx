import React, { useState } from "react";
import { Button, ThemeProvider } from "fictoan-react";
import { CompanyBotWidgetStyled } from "./CompanyBotWidget.styled";
import { ChatBot } from "./assets/icons/ChatBot";
import CompanyBotWidgetModal from "./CompanyBotWidgetModal";
import { CompanyBotLightTheme } from "./assets/CompanyBot.light.theme";

export interface SetuBotWidgetProps {
    description?: string;
    heading?: string;
    primaryColor?: string;
    projectKey: string;
}

export function SetuBotWidget({ primaryColor, ...props }: SetuBotWidgetProps) {
    const [openModal, setOpenModal] = useState(false);

    const closeModal = () => {
        setOpenModal(false);
    };
    return (
        <ThemeProvider theme={CompanyBotLightTheme}>
            <CompanyBotWidgetStyled>
                <div className="feedback-widget">
                    <Button
                        className="feedback-widget-trigger-button"
                        size="small"
                        shadow="hard"
                        onClick={() => setOpenModal(true)}
                    >
                        <ChatBot />
                    </Button>
                    {openModal ? (
                        <CompanyBotWidgetModal
                            closeModal={closeModal}
                            {...props}
                        />
                    ) : null}
                </div>
            </CompanyBotWidgetStyled>
        </ThemeProvider>
    );
}
