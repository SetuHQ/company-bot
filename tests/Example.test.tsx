import React from 'react';
import renderer from 'react-test-renderer';
import { Example } from '../src';

it('renders correctly', () => {
  const tree = renderer
    .create(<Example text="Clicked this many times" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
