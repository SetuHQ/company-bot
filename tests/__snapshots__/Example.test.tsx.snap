// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<button
  onClick={[Function]}
  type="button"
>
  Clicked this many times 0
</button>
`;
